gh-112334: Regression test that vfork is used when expected.

gpshead opened the pull request:
Adds a regression test #112334. This is the regression test I was pondering in #112617.

This also ensures strace is present in our CI GH (It may have already been there, ubuntu-standard includes it, now we're explicit about the need). I also made sure to install it on some Linux buildbots that I maintain. From my testing, some clearly do not have it.

If this test ever proves a challenge to maintain, further restricting it to a specific platform is fine. The way it works depends on details that can vary between different Linux platforms. ex: aarch64 uses more modern syscalls than x86_64 for the same task.

<!-- gh-issue-number: gh-112334 -->
* Issue: gh-112334
<!-- /gh-issue-number -->

gpshead committed:
gh-112334: Regression test that vfork is used when expected.

gpshead committed:
Move the vfork tests to POSIXProcessTestCase.

This way they only get run once instead of twice via NoPoll.

gpshead committed:
Add strace to the Linux CI images.

gpshead committed:
fix the strace availability check.

gpshead committed:
"really" fix the strace detection.

gpshead committed:
consistent use of dedent

gpshead committed:
Loosen the test, different architectures use different syscall variants.

gpshead committed:
non-late-night regex fixing.

serhiy-storchaka commented on Lib/test/support/script_helper.py:
```diff
@@ -92,13 +92,28 @@ def fail(self, cmd_line):
+    run it.
+    """
     env_required = interpreter_requires_environment()
+    run_using_command = env_vars.pop('_run_using_command', None)
```
Perhaps use a double underscore for consistency with other arguments (`__cwd`, `__isolated`)?

serhiy-storchaka commented on Lib/test/test_subprocess.py:
```diff
@@ -3360,6 +3345,89 @@ def exit_handler():
+                    _run_using_command=strace_command,
+                )
+                # Ensure neither vfork() or clone(..., flags=...|CLONE_VFORK|...).
+                self.assertNotRegex(non_vfork_result.err, br"(?mi)vfork")
```
Is `?m` needed here?

serhiy-storchaka commented on Lib/test/test_subprocess.py:
```diff
@@ -3360,6 +3345,89 @@ def exit_handler():
+                    _run_using_command=strace_command,
+            )
+            # Match both vfork() and clone(..., flags=...|CLONE_VFORK|...)
+            self.assertRegex(vfork_result.err, br"(?mi)vfork")
```
Is `?m` needed here?

serhiy-storchaka reviewed (approved):
LGTM.

gpshead committed:
Address code review comments.

gpshead committed:
simplify the syscalls strace reports.

This makes output easier to understand if debugging a failure.  (removes
execve which shows up for the initial process launch by strace on _some_
platforms, adds clone2 which claims to only be used on an obsolete
architecture but if it cropped up again for some future reason would be
annoying to have break the test)

gpshead committed:
Merge branch 'main' into test/subprocess/112334
